### PR TITLE
[TAS-488] ✨ Deduct the amount of Stripe tx fee for NFT checked-out with LIKE

### PR DESCRIPTION
**Pull request by WeiJun0827**

### CODE REVIEW

File: config/config.js [hunk 1]
```diff
@@ -53,6 +53,7 @@ config.LIKER_NFT_MINT_MIN_BATCH = 7; // limit for USD base
 config.LIKER_NFT_LIKE_TO_USD_CONVERT_RATIO = 1024;
 config.LIKER_NFT_MIN_USD_PRICE = 1;
 config.LIKER_NFT_GAS_FEE = '200000';
+config.LIKER_NFT_PAY_BY_LIKE_PRICE_OFF = 0.4;
```
Comment (nwingt) [inline config/config.js]:
```suggestion
config.LIKER_NFT_LIKE_PRICE_DISCOUNT_IN_USD = 0.4;
```

File: src/routes/likernft/purchase.ts [hunk 2]
```diff
@@ -93,6 +98,9 @@ router.post(
         likeWallet = wallet as string;
         grantedAmount = 0;
       }
+      const priceModifier = (p: BigNumber) => p
+        .minus(LIKER_NFT_TX_FEE_DISCOUNT_IN_USD)
+        .multipliedBy(1 - LIKER_NFT_LIKE_PRICE_DISCOUNT_RATIO);
```
Comment (nwingt) [inline src/routes/likernft/purchase.ts]: can we refactor this function to util and reuse in `calculatePayment()`

File: src/util/api/likernft/fiat/index.ts [hunk 3]
```diff
@@ -37,15 +39,28 @@ export async function getPurchaseInfoList(iscnPrefixes, classIds) {
...
+      const discountedPrice = new BigNumber(price)
+        .minus(LIKER_NFT_TX_FEE_DISCOUNT_IN_USD)
+        .multipliedBy(1 - LIKER_NFT_LIKE_PRICE_DISCOUNT_RATIO);
+      return discountedPrice.isLessThanOrEqualTo(0)
+        ? acc
+        : acc.plus(discountedPrice);
+    }, new BigNumber(0))
+    .dividedBy(rate)
+    .multipliedBy(1 + buffer)
+    .toFixed(0, BigNumber.ROUND_UP));
```
Comment (nwingt) [inline src/util/api/likernft/fiat/index.ts]: How about?
```js
const priceReducer = (acc, { price }) => {
  const discountedPrice = new BigNumber(price)
    .minus(LIKER_NFT_TX_FEE_DISCOUNT_IN_USD)
    .multipliedBy(1 - LIKER_NFT_LIKE_PRICE_DISCOUNT_RATIO);
  return discountedPrice.isLessThanOrEqualTo(0)
    ? acc
    : acc.plus(discountedPrice);
};
const priceReducerWithoutDiscount = (acc, { price }) => acc.plus(price);

const calculateTotalLIKEPrice = reducer => Number(
  purchaseInfoList
    .reduce(reducer, new BigNumber(0))
    .dividedBy(rate)
    .multipliedBy(1 + buffer)
    .toFixed(0, BigNumber.ROUND_UP)
);

const totalLIKEPrice = calculateTotalLIKEPrice(priceReducer);
const totalLIKEPriceNoDiscount = calculateTotalLIKEPrice(priceReducerWithoutDiscount);
```

Comment (WeiJun0827) [reply src/util/api/likernft/fiat/index.ts]: I extracted the modifier and used the same reducer for both cases, to guard exceptional/unwanted cases.

File: config/config.js [hunk 4]
```diff
@@ -53,6 +53,8 @@ config.LIKER_NFT_MINT_MIN_BATCH = 7; // limit for USD base
 config.LIKER_NFT_LIKE_TO_USD_CONVERT_RATIO = 1024;
 config.LIKER_NFT_MIN_USD_PRICE = 1;
 config.LIKER_NFT_GAS_FEE = '200000';
+config.LIKER_NFT_TX_FEE_DISCOUNT_IN_USD = 0.5;
+config.LIKER_NFT_LIKE_PRICE_DISCOUNT_RATIO = 0.05;
```
Comment (williamchong) [inline config/config.js]: Should not hardcode LIKE discount as conflict, we should just calculate fiat price - stripe fee and use that as discounted price

**Review by williamchong — CHANGES_REQUESTED** [review]

File: src/util/api/likernft/fiat/index.ts [hunk 5]
```diff
@@ -37,15 +40,29 @@ export async function getPurchaseInfoList(iscnPrefixes, classIds) {
...
+      .dividedBy(rate)
+      .multipliedBy(1 + buffer)
+      .toFixed(0, BigNumber.ROUND_UP),
+  );
+  const priceModifierWithoutDiscount = (price: BigNumber) => price;
+
+  const totalLIKEPrice = calculateTotalLIKEPrice(priceModifierForCheckoutWithLIKE);
+  const totalLIKEPriceNoDiscount = calculateTotalLIKEPrice(priceModifierWithoutDiscount);
   const totalFiatBigNum = purchaseInfoList
     .reduce((acc, { price }) => acc.plus(price), new BigNumber(0));
```
Comment (nwingt) [inline src/util/api/likernft/fiat/index.ts]: I wonder why `purchaseInfoList` is `{ price }[]` here but `price[]` in above

Comment (WeiJun0827) [reply src/util/api/likernft/fiat/index.ts]: Indeed, should add a `map` before calling reducer.

File: src/routes/likernft/payment.ts [hunk 6]
```diff
@@ -11,8 +11,13 @@ router.get(
     try {
       const { iscnPrefixes, classIds } = res.locals;
       const purchaseInfoList = await getPurchaseInfoList(iscnPrefixes, classIds);
-      const { totalLIKEPrice, totalFiatPriceString } = await calculatePayment(purchaseInfoList);
+      const {
+        totalLIKEPriceNoDiscount,
```
Comment (williamchong) [inline src/routes/likernft/payment.ts]: How about prediscount instead of nodiscount

File: src/util/api/likernft/purchase.ts [hunk 7]
```diff
@@ -639,6 +641,21 @@ async function updateDocsForMissingSoldNFT(t, {
   });
 }
 
+// NOTE: Stripe fee per order is 5.4% + 30 cents
+export function rewardModifierForCheckoutWithUSD(priceInUSD: BigNumber, nftCountOfOrder: number) {
+  return priceInUSD
+    // all NFT share flat fee
+    .minus(new BigNumber(LIKER_NFT_STRIPE_FEE_USD_INTERCEPT).dividedBy(nftCountOfOrder))
```
Comment (williamchong) [inline src/util/api/likernft/purchase.ts]: For consistency we should just minus the flat price for all nft?

**Review by williamchong — APPROVED** [review]

**Pull request merged**